File: testdata/fixtures/go-sample/auth/handler.go
// Package auth provides JWT token validation and claims extraction.
package auth

import (
	"errors"
	"fmt"
	"strings"
	"time"
)

const (
	// bearerPrefix is the expected prefix on Authorization header values.
	bearerPrefix = "Bearer "

	// tokenTTL is the maximum token lifetime.
	tokenTTL = 24 * time.Hour
)

// AuthError represents an authentication failure with a machine-readable code.
type AuthError struct {
	Message string
	Code    string
}

func (e *AuthError) Error() string {
	return fmt.Sprintf("auth error [%s]: %s", e.Code, e.Message)
}

// Sentinel errors for common authentication failures.
var (
	ErrMalformedToken = errors.New("malformed token")
	ErrTokenExpired   = errors.New("token expired")
	ErrInvalidSig     = errors.New("invalid signature")
)

// Claims holds the decoded JWT claims extracted from a validated token.
type Claims struct {
	// Sub is the subject (user ID).
	Sub string
	// Role is the user's assigned role.
	Role string
	// Exp is the expiration time as a Unix timestamp.
	Exp int64
	// Iat is the issued-at time as a Unix timestamp.
	Iat int64
	// Issuer identifies the token issuer.
	Issuer string
}

// IsExpired reports whether the token has expired.
func (c *Claims) IsExpired() bool {
	return time.Now().Unix() > c.Exp
}

// RemainingDuration returns the time until the token expires.
func (c *Claims) RemainingDuration() time.Duration {
	return time.Until(time.Unix(c.Exp, 0))
}

// AuthHandler validates tokens and enforces access control.
type AuthHandler struct {
	secret []byte
}

// NewAuthHandler creates a handler with the given HMAC secret.
func NewAuthHandler(secret string) *AuthHandler {
	return &AuthHandler{secret: []byte(secret)}
}

// ValidateToken parses and validates a bearer token from the Authorization header.
func (h *AuthHandler) ValidateToken(authHeader string) (*Claims, error) {
	return ValidateToken(authHeader, h.secret)
}

// ValidateToken is a package-level function that validates a bearer token.
func ValidateToken(authHeader string, secret []byte) (*Claims, error) {
	if !strings.HasPrefix(authHeader, bearerPrefix) {
		return nil, &AuthError{
			Message: "missing Bearer prefix",
			Code:    "MALFORMED",
		}
	}

	token := strings.TrimPrefix(authHeader, bearerPrefix)
	parts := strings.Split(token, ".")

	if len(parts) != 3 {
		return nil, &AuthError{
			Message: fmt.Sprintf("expected 3 parts, got %d", len(parts)),
			Code:    "MALFORMED",
		}
	}

	// In a real implementation, verify the HMAC signature here.
	_ = parts[0] // header
	_ = parts[1] // payload
	_ = parts[2] // signature
	_ = secret

	now := time.Now()
	claims := &Claims{
		Sub:    "user-1",
		Role:   "user",
		Exp:    now.Add(tokenTTL).Unix(),
		Iat:    now.Unix(),
		Issuer: "cruxe",
	}

	if claims.IsExpired() {
		return nil, &AuthError{
			Message: "token has expired",
			Code:    "EXPIRED",
		}
	}

	return claims, nil
}

// RequireRole checks that the claims contain at least the given role level.
func RequireRole(claims *Claims, minimum string) error {
	roleOrder := map[string]int{
		"guest":     0,
		"user":      1,
		"moderator": 2,
		"admin":     3,
	}

	actual, ok := roleOrder[claims.Role]
	if !ok {
		return fmt.Errorf("unknown role: %s", claims.Role)
	}

	required, ok := roleOrder[minimum]
	if !ok {
		return fmt.Errorf("unknown required role: %s", minimum)
	}

	if actual < required {
		return &AuthError{
			Message: fmt.Sprintf("requires %s role, user has %s", minimum, claims.Role),
			Code:    "FORBIDDEN",
		}
	}

	return nil
}

File: testdata/fixtures/ts-sample/src/auth/handler.ts
/**
 * Authentication handler for validating JWT tokens and managing sessions.
 */

import { User, UserRole } from "../models/user";

/** Prefix expected on Authorization header values. */
const BEARER_PREFIX = "Bearer ";

/** Token expiration time in milliseconds (24 hours). */
const TOKEN_TTL_MS = 24 * 60 * 60 * 1000;

/** Decoded JWT claims. */
export interface TokenClaims {
  sub: string;
  role: UserRole;
  exp: number;
  iat: number;
  iss: string;
}

/** Errors thrown by authentication operations. */
export class AuthError extends Error {
  constructor(
    message: string,
    public readonly code:
      | "MALFORMED"
      | "EXPIRED"
      | "INVALID_SIGNATURE"
      | "FORBIDDEN",
  ) {
    super(message);
    this.name = "AuthError";
  }
}

/**
 * Validate a bearer token and return the decoded claims.
 *
 * @param authHeader - The full Authorization header value.
 * @param secret - The HMAC secret for signature verification.
 * @returns Decoded token claims.
 * @throws {AuthError} If the token is invalid or expired.
 */
export function validateToken(authHeader: string, secret: string): TokenClaims {
  if (!authHeader.startsWith(BEARER_PREFIX)) {
    throw new AuthError("Missing Bearer prefix", "MALFORMED");
  }

  const token = authHeader.slice(BEARER_PREFIX.length);
  const parts = token.split(".");

  if (parts.length !== 3) {
    throw new AuthError(`Expected 3 parts, got ${parts.length}`, "MALFORMED");
  }

  // In a real implementation, verify the HMAC signature here.
  const _header = parts[0];
  const _payload = parts[1];
  const _signature = parts[2];
  void secret;

  const claims: TokenClaims = {
    sub: "user-1",
    role: UserRole.User,
    exp: Date.now() + TOKEN_TTL_MS,
    iat: Date.now(),
    iss: "cruxe",
  };

  if (claims.exp < Date.now()) {
    throw new AuthError("Token has expired", "EXPIRED");
  }

  return claims;
}

/**
 * Handles authentication for incoming HTTP requests.
 */
export class AuthHandler {
  private readonly secret: string;

  constructor(secret: string) {
    this.secret = secret;
  }

  /** Authenticate a request and return the associated user. */
  async authenticate(headers: Record<string, string>): Promise<User> {
    const authHeader = headers["authorization"];
    if (!authHeader) {
      throw new AuthError("Missing Authorization header", "MALFORMED");
    }

    const claims = validateToken(authHeader, this.secret);

    return {
      id: claims.sub,
      username: `user-${claims.sub}`,
      email: `${claims.sub}@example.com`,
      role: claims.role,
      active: true,
      createdAt: new Date(claims.iat),
    };
  }

  /** Check whether a user has the minimum required role. */
  requireRole(user: User, minimum: UserRole): void {
    const roleOrder = [
      UserRole.Guest,
      UserRole.User,
      UserRole.Moderator,
      UserRole.Admin,
    ];
    if (roleOrder.indexOf(user.role) < roleOrder.indexOf(minimum)) {
      throw new AuthError(
        `Requires ${minimum} role, user has ${user.role}`,
        "FORBIDDEN",
      );
    }
  }
}
